### PR TITLE
Preference rank

### PR DESCRIPTION
**Pull request by appleJax**
Adds trophy to top 3 ranked timeslots. Shows best timeslot on event card on the `/events` route

screen shot 2017-10-28 at 2 39 50 pm
screen shot 2017-10-28 at 2 39 36 pm

### CODE REVIEW

**Comment by gabrielbonner** [comment]
@kylestengline Want to take a look at this whenever you have time?  Kevin and I worked on it so it would be better to have a fresh set of eyes on it in case we overlooked something...  :)

@jeremyjung ^ (if you're not on vacation)

**Comment by kylestengline** [comment]
@appleJax and @gabrielbonner looks good to me! Can we keep the `FU hound` commit? haha

File: app/javascript/components/TimeslotRow.jsx [hunk 1]
```diff
@@ -4,6 +4,16 @@ import { formatTimeslot } from '../utils/dateFormat'
 import Preference from './Preference'
 
 const TimeSlotRow = ({ timeslot, duration, confirmModal, updatePreference }) => {
+  let rankColor = 'gold'
```
Comment (jeremyjung) [inline app/javascript/components/TimeslotRow.jsx]: I find the switch statement more readable when changing behavior based on a single variable

```javascript
  let rankColor
  let tooltipText

  switch (timeslot.rank) {
    case 1:
      rankColor = 'gold'
      tooltipText = <div>Best<br/>Timeslot</div>
      break
    case 2:
      rankColor = 'silver'
      tooltipText = '2nd Best Timeslot'
      break
    case 3:
      rankColor = 'bronze'
      tooltipText = '3rd Best Timeslot'
      break
  }
```

File: app/models/event.rb [hunk 2]
```diff
@@ -19,4 +19,53 @@ class Event < ApplicationRecord
...
+      if i < 3
+        ranking = 1 + i
+        timeslot.define_singleton_method(:rank) { ranking }
+      end
+      timeslot
+    end
+  end
+
+  def preference_score(preference)
+    return 1 if preference == 'tentative'
```
Comment (jeremyjung) [inline app/models/event.rb]: I think a switch statement is also more readable here.  Typically you use implicit returns in ruby unless you are exiting a function early.
```ruby
    case preference
    when 'tentative' then 1
    when 'available' then 2
    when 'preferred' then 3
    when 'not_available' then -100
    else 0
    end
```

File: app/models/event.rb [hunk 3]
```diff
@@ -19,4 +19,53 @@ class Event < ApplicationRecord
...
+    timeslots_with_score = add_score_to_timeslot(timeslots)
+
+    sorted_timeslots = timeslots_with_score.sort do |a, b|
+      b.score <=> a.score
+    end
+
+    add_rank_to_timeslots sorted_timeslots
+  end
+
+  def add_score_to_timeslot(timeslots)
```
Comment (jeremyjung) [inline app/models/event.rb]: Defining singleton methods in-line and having multiple implementations looks pretty spooky to me 🎃.  I need to spend more time reviewing this to understand what is going on.

File: app/javascript/components/TimeslotRow.jsx (hunk 1 above)
Comment (appleJax) [reply app/javascript/components/TimeslotRow.jsx]: Good call, fixed

File: app/models/event.rb (hunk 2 above)
Comment (appleJax) [reply app/models/event.rb]: 👍

File: app/models/event.rb (hunk 3 above)
Comment (appleJax) [reply app/models/event.rb]: @jeremyjung I refactored by adding a `rank` method to the `Timeslot` model that always returns nil. Now, in the `Event` model, we only override the `rank` method for the 3 best timeslots. Not sure how else to add ranking info to a timeslot instance.

**Comment by jeremyjung** [comment]
I refactored the event's best_timeslot runtime method definition.  I think it should be possible to clean up the timeslot code too.

**Comment by jeremyjung** [comment]
Warden test helpers are used for [signing in as a user inside of capybara tests,](https://github.com/plataformatec/devise/wiki/How-To:-Test-with-Capybara) but actually I did not intend to check that in as a part of this commit.  I guess I accidentally left that in from earlier work.  You can remove it.

If you think we will always include the timeslots then we don't need a separate serializer.

**Comment by kylestengline** [comment]
Hey all,
I was looking over this again and wanted to check if it's ok to merge, or does this need further review?

**Comment by appleJax** [comment]
I think it's good to go. We can always put in another PR if we see a problem later.

**Pull request merged**